tools/tcat_ble_client/cli/base_commands.py

```diff
@@ -159,10 +169,111 @@ def get_log_string(self) -> str:
+        elif type == "pskc":
+            code = bytes.fromhex(args[1])
+            tlv_type = TcatTLVType.PRESENT_PSKC_HASH.value
+        elif type == "install":
+            code = bytes(args[1], 'utf-8')
+            tlv_type = TcatTLVType.PRESENT_INSTALL_CODE_HASH.value
+        else:
+            raise DataNotPrepared("Hash code name incorrect.")
+        bless: BleStreamSecure = context['ble_sstream']
+        if bless.peer_public_key is None:
+            raise DataNotPrepared("Peer certificate not present.")
+
+        if bless.peer_challenge is None:
+            raise DataNotPrepared("Peer challenge not present.")
+
+        hash = hmac.new(code, digestmod=sha256)
+        hash.update(bless.peer_challenge)
+        hash.update(bless.peer_public_key)
+
+        data = TLV(tlv_type, hash.digest()).to_bytes()
+        return data
+
+
+class GetPskdHash(Command):
+
+    def get_log_string(self) -> str:
+        return 'Retrieving peer PSKd hash.'
+
+    def get_help_string(self) -> str:
+        return 'Get calculated PSKd hash.'
+
+    async def execute_default(self, args, context):
+        bless: BleStreamSecure = context['ble_sstream']
+
+        print(self.get_log_string())
+        try:
+            if bless.peer_public_key is None:
+                print("Peer certificate not present.")
+                return
+            challenge_size = 8
+            challenge = token_bytes(challenge_size)
+            pskd = bytes(args[0], 'utf-8')
+            data = TLV(TcatTLVType.GET_PSKD_HASH.value, challenge).to_bytes()
+            response = await bless.send_with_resp(data)
+            if not response:
+                return
+            tlv_response = TLV.from_bytes(response)
+            if tlv_response.value != None:
+                hash = hmac.new(pskd, digestmod=sha256)
+                hash.update(challenge)
+                hash.update(bless.peer_public_key)
+                digest = hash.digest()
+                if digest == tlv_response.value:
+                    print('Requested hash is valid.')
+                else:
+                    print('Requested hash is NOT valid.')
+            return CommandResultTLV(tlv_response)
+        except DataNotPrepared as err:
+            print('Command failed', err)
+
+
+class GetRandomNumberChallenge(Command):
+
+    def get_log_string(self) -> str:
+        return 'Retrieving random challenge.'
+
+    def get_help_string(self) -> str:
+        return 'Get the device random number challenge.'
+
+    async def execute_default(self, args, context):
+        bless: BleStreamSecure = context['ble_sstream']
+
+        print(self.get_log_string())
+        try:
+            data = TLV(TcatTLVType.GET_RANDOM_NUMBER_CHALLENGE.value, bytes()).to_bytes()
+            response = await bless.send_with_resp(data)
+            if not response:
+                return
+            tlv_response = TLV.from_bytes(response)
+            if tlv_response.value != None:
```

Comment:
One more thing - here we could add a size check on the random challenge. This can be useful to catch non-conforming DUTs early on. Otherwise, a DUT sending 1 byte might even pass the tests.

Reply:
@EskoDijk i have updated the code.